src/frogger.cpp

```diff
@@ -31,40 +31,86 @@ int get_hiv_steps_per_year(SEXP r_hiv_steps_per_year) {
   int age_groups_hiv;
-  SEXP data;
   if (hiv_age_stratification == "full") {
     age_groups_hiv = 66;
-    data = projection_parameters["hAG_SPAN_full"];
   } else if (hiv_age_stratification == "coarse") {
     age_groups_hiv = 9;
-    data = projection_parameters["hAG_SPAN_coarse"];
   } else {
     Rcpp::stop(
         "Invalid HIV age stratification must be 'full' or 'coarse' got '%s'.",
         hiv_age_stratification);
   }
-  if (LENGTH(data) != age_groups_hiv) {
-    Rcpp::stop("Invalid size of data, expected %d got %d", age_groups_hiv,
-               LENGTH(data));
+  return age_groups_hiv;
+}
+
+template<typename... Args>
+auto parse_data_int(const Rcpp::List data, const std::string& key, Args... dims) {
+  constexpr std::size_t rank = sizeof...(dims);
+  Eigen::array<int, rank> dimensions{ static_cast<int>(dims)... };
+
+  int product = 1;
+  for (size_t i = 0; i < rank; ++i) {
+    product *= dimensions[i];
   }
-  leapfrog::TensorMap1<int> age_groups_hiv_span(INTEGER(data), age_groups_hiv);
-  return age_groups_hiv_span;
+  SEXP array_data = data[key];
+  // In cases where the input data has project years we might not use all of it model fit
+  // So we can take create a Map over a smaller slice of the data
+  // As long as this is true we can be confident we're not referencing invalid memory
+  if (LENGTH(array_data) < product) {
+    Rcpp::stop("Invalid size of data for '%s', expected %d got %d",
+               key,
+               product,
+               LENGTH(array_data));
+  }
+
+  return Eigen::TensorMap<Eigen::Tensor<int, rank>>(INTEGER(array_data), static_cast<int>(dims)...);
+}
+
+template<typename... Args>
+auto parse_data_double(const Rcpp::List data, const std::string& key, Args... dims) {
+  constexpr std::size_t rank = sizeof...(dims);
+  Eigen::array<int, rank> dimensions{ static_cast<int>(dims)... };
+
+  int product = 1;
+  for (size_t i = 0; i < rank; ++i) {
+    product *= dimensions[i];
+  }
+  SEXP array_data = data[key];
+  // In cases where the input data has project years we might not use all of it model fit
+  // So we can take create a Map over a smaller slice of the data
+  // As long as this is true we can be confident we're not referencing invalid memory
+  if (LENGTH(array_data) < product) {
+    Rcpp::stop("Invalid size of data for '%s', expected %d got %d.",
+               key,
+               product,
+               LENGTH(array_data));
+  }
+
+  return Eigen::TensorMap<Eigen::Tensor<double, rank>>(REAL(array_data), static_cast<int>(dims)...);
 }
 
 std::vector<int> parse_output_steps(Rcpp::NumericVector output_steps, int proj_years) {
   return Rcpp::as<std::vector<int>>(output_steps);
 }
 
+void validate_stratification(std::string stratification) {
+  if (!(stratification == "full" || stratification == "coarse")) {
+    Rcpp::stop(
+        "Invalid HIV age stratification must be 'full' or 'coarse' got '%s'.",
+        stratification);
+  }
+}
```

Comment:
I also toyed around with validating and returning an enum but then where I want to use this I want to append `_full` or `_coarse` on the end of the string so felt a bit weird to convert from the string to an enum and back again to a string again so I removed that